numpy/core/src/multiarray/convert_datatype.c

```diff
@@ -629,35 +629,43 @@ type_num_unsigned_to_signed(int type_num)
 /*
  * Compare two field dictionaries for castability.
  *
- * Return 1 if 'field1' can be cast to 'field2' according to the rule
+ * Return 1 if 'from' can be cast to 'to' according to the rule
  * 'casting', 0 if not.
  *
- * Castabiliy of field dictionaries is defined recursively: 'field1' and
- * 'field2' must have the same field names (possibly in different
- * orders), and the corresponding field types must be castable according
- * to the given casting rule.
+ * Castabiliy of field dictionaries is defined recursively: 'to' and
+ * 'from' must have the same field names (in the same order), and the
+ * corresponding field types must be castable according to the given casting
+ * rule.
  */
 static int
-can_cast_fields(PyObject *field1, PyObject *field2, NPY_CASTING casting)
+can_cast_fields(PyArray_Descr *from, PyArray_Descr *to, NPY_CASTING casting)
 {
     Py_ssize_t ppos;
     PyObject *key;
     PyObject *tuple1, *tuple2;
+    int val;
 
-    if (field1 == field2) {
+    if (from->fields == to->fields) {
         return 1;
     }
-    if (field1 == NULL || field2 == NULL) {
+    if (from->fields == NULL || to->fields == NULL) {
         return 0;
     }
-    if (PyDict_Size(field1) != PyDict_Size(field2)) {
+
+    val = PyObject_RichCompareBool(from->names, to->names, Py_EQ);
+    if (val != 1 || PyErr_Occurred()) {
+        PyErr_Clear();
```

Comment:
I don;t think you should clear the error here, else `MemoryError` translates to "can't cast" which is nonsense.